src/stream.h

```diff
@@ -17,82 +17,142 @@
+  void WriteU8(uint32_t value,
+               const char* desc = nullptr,
+               PrintChars print_chars = PrintChars::No) {
+    assert(value <= UINT8_MAX);
+    Write(static_cast<uint8_t>(value), desc, print_chars);
+  }
+  void WriteU32(uint32_t value,
+                const char* desc = nullptr,
+                PrintChars print_chars = PrintChars::No) {
+    Write(value, desc, print_chars);
+  }
+  void WriteU64(uint64_t value,
+                const char* desc = nullptr,
+                PrintChars print_chars = PrintChars::No) {
+    Write(value, desc, print_chars);
+  }
+  void WriteChar(char c,
+                 const char* desc = nullptr,
+                 PrintChars print_chars = PrintChars::No) {
+    WriteU8(c, desc, print_chars);
+  }
+
+  // Dump memory as text, similar to the xxd format.
+  void WriteMemoryDump(const void* start,
                        size_t size,
-                       size_t offset,
-                       PrintChars print_chars,
-                       const char* prefix,
-                       const char* desc);
-
-static WABT_INLINE void write_output_buffer_memory_dump(
-    Stream* stream,
-    struct OutputBuffer* buf) {
-  write_memory_dump(stream, buf->start, buf->size, 0, PrintChars::No, nullptr,
-                    nullptr);
-}
-
-/* Helper function for writing enums as u8. */
-template <typename T>
-void write_u8_enum(Stream* stream, T value, const char* desc) {
-  write_u8(stream, static_cast<uint32_t>(value), desc);
-}
+                       size_t offset = 0,
+                       const char* prefix = nullptr,
+                       const char* desc = nullptr,
+                       PrintChars print_chars = PrintChars::No);
+
+  // Convenience functions for writing enums.
+  template <typename T>
+  void WriteU8Enum(T value,
+                   const char* desc = nullptr,
+                   PrintChars print_chars = PrintChars::No) {
+    WriteU8(static_cast<uint32_t>(value), desc, print_chars);
+  }
+
+ private:
+  template <typename T>
+  void Write(const T& data, const char* desc, PrintChars print_chars) {
+    WriteData(&data, sizeof(data), desc, print_chars);
+  }
+
+  Writer* writer_;  // Not owned.
+  size_t offset_;
+  Result result_;
+  // Not owned. If non-null, log all writes to this stream.
+  Stream* log_stream_;
+};
+
+class MemoryStream : public Stream {
+ public:
+  MemoryStream();
+
+  MemoryWriter& writer() { return writer_; }
+
+  std::unique_ptr<OutputBuffer> ReleaseOutputBuffer() {
+    return writer_.ReleaseOutputBuffer();
+  }
+
+  Result WriteToFile(const char* filename) {
+    return writer_.output_buffer().WriteToFile(filename);
+  }
+
+ private:
+  MemoryWriter writer_;
+};
+
+class FileStream : public Stream {
+ public:
+  explicit FileStream(const char* filename);
+  explicit FileStream(FILE*);
+
+  FileWriter& writer() { return writer_; }
+
+  static std::unique_ptr<FileStream> CreateStdout();
+  static std::unique_ptr<FileStream> CreateStderr();
```

Comment:
Could do, I guess. This way the names are scoped, which is nice. Also, seems reasonable to have constructor functions be defined as static members.